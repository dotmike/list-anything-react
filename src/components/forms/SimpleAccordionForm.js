import React, { Component } from "react";
import PropTypes from "prop-types";
import { Accordion, Form, Icon } from "semantic-ui-react";

class SimpleAccordionForm extends Component {
  state = {
    active: true,
    errors: {}
  };

  onAccordionClick = () =>
    this.setState(prevState => ({ active: !prevState.active }));

  onSubmit = e => {
    e.preventDefault();

    const data = this.props.getData(this.getNodeByName);
    const errors = this.props.validate(data);

    this.setState({ errors });
    if (Object.keys(errors).length === 0) this.props.submit(data);

    this.props.resetFormData(this.getNodeByName);
    this.focusOnFirstInput();
  };

  getNodeByName = (name, nodeType = "input") =>
    document.querySelector(`#${this.props.id} ${nodeType}[name="${name}"]`);

  focusOnFirstInput = () =>
    document.querySelector(`#${this.props.id} input[name]`).focus();

  render() {
    const { id, title, render } = this.props;
    const { active } = this.state;
    return (
      <Accordion id={id} as={Form.Field}>
        <Accordion.Title
          name="tabs"
          onClick={this.onAccordionClick}
          active={active}
        >
          <Icon name="dropdown" />
          {title}
        </Accordion.Title>
        <Accordion.Content active={active}>
          {render(this.state, this.onSubmit)}
        </Accordion.Content>
      </Accordion>
    );
  }
}

SimpleAccordionForm.propTypes = {
  // ownProps
  id: PropTypes.string,
  title: PropTypes.string,
  validate: PropTypes.func,
  render: PropTypes.func.isRequired,
  submit: PropTypes.func.isRequired,
  getData: PropTypes.func.isRequired,
  resetFormData: PropTypes.func.isRequired
};

SimpleAccordionForm.defaultProps = {
  id: `accordion-form-${Math.floor(Math.random() * 100000)}`,
  title: "Accordion",
  validate: () => ({})
};

export default SimpleAccordionForm;
