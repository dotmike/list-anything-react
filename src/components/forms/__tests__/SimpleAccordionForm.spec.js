import React from "react";
import { shallow } from "enzyme";

import SimpleAccordionForm from "../SimpleAccordionForm";

const defaultProps = {
  id: "accordion-form-test",
  title: "Test",
  validate: () => ({}),
  render: () => <div>Testando...</div>,
  submit: () => {},
  getData: () => ({}),
  resetFormData: () => {}
};

describe("SimpleAccordionForm", () => {
  it("renders correctly", () => {
    const wrapper = shallow(<SimpleAccordionForm {...defaultProps} />);
    expect(wrapper).toMatchSnapshot();
  });
});
