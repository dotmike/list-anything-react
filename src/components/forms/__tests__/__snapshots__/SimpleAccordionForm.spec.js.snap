// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleAccordionForm renders correctly 1`] = `
<Accordion
  as={[Function]}
  id="accordion-form-test"
>
  <AccordionTitle
    active={true}
    name="tabs"
    onClick={[Function]}
  >
    <Icon
      as="i"
      name="dropdown"
    />
    Test
  </AccordionTitle>
  <AccordionContent
    active={true}
  >
    <div>
      Testando...
    </div>
  </AccordionContent>
</Accordion>
`;
